anomaly/ordered_tree_editing_distance.cpp

```diff
@@ -0,0 +1,145 @@
+unsigned int OrderedTreeEditingDistance::get_distance()
+{
+    if (!_computed)
+    {
+        _result = _zhang_shasha_algorithm();
+        _computed = true;
+    }
+    return _result;
+}
+
+static int _precompute(std::shared_ptr<TreeNode> node,
+                       std::vector<std::string>& post_order,
+                       std::vector<int>& left_most_leaf,
+                       std::vector<int>& lr_key_roots)
+{
+    int left_most_child_pos;
+    for (auto child = node->children.begin(); child != node->children.end(); ++child)
+    {
+        if (child == node->children.begin()) 
+        {
+            left_most_child_pos = _precompute(child->first, post_order, left_most_leaf, lr_key_roots);
+        }
+        else
+        {
+            int child_pos = _precompute(child->first, post_order, left_most_leaf, lr_key_roots);
+            lr_key_roots.push_back(child_pos);
+        }
+    }
+
+    post_order.push_back(node->data.label);
+    if (node->children.size() == 0)
+    {
+        left_most_leaf.push_back(post_order.size() - 1);
+    }
+    else
+    {
+        left_most_leaf.push_back(left_most_leaf[left_most_child_pos]);
+    }
+    return post_order.size() - 1;
+}
+
+unsigned int OrderedTreeEditingDistance::_tree_dist(int src_node, 
+                                                    int dst_node,
+                                                    std::vector<std::vector<unsigned int>>& treedist,
+                                                    std::vector<std::vector<bool>>& computed)
+{
+    std::vector<std::vector<unsigned int>> forestdist(src_node + 2, std::vector<unsigned int>(dst_node + 2));
+
+    forestdist[0][0] = 0;
+    for (size_t i = _left_most_leaf_src[src_node]; i <= src_node; ++i)
+    {
+        forestdist[i + 1][0] = forestdist[i][0] + _config->get_delete_cost();
+    }
+
+    for (size_t j = _left_most_leaf_dst[dst_node]; j <= dst_node; ++j)
+    {
+        forestdist[0][j + 1] = forestdist[0][j] + _config->get_insert_cost();
+    }
+
+    for (size_t i = _left_most_leaf_src[src_node]; i <= src_node; ++i)
+    {
+        for (size_t j = _left_most_leaf_dst[dst_node]; j <= dst_node; ++j)
+        {
+            int prev_i = _left_most_leaf_src[src_node] <= i - 1 ? i - 1 : -1;
+            int prev_j = _left_most_leaf_dst[dst_node] <= j - 1 ? j - 1 : -1;
+            if (_left_most_leaf_src[i] == _left_most_leaf_src[src_node] &&
+                _left_most_leaf_dst[j] == _left_most_leaf_dst[dst_node])
+            {
+
+                forestdist[i + 1][j + 1] = std::min(
+                        forestdist[prev_i + 1][j + 1] + _config->get_delete_cost(),
+               std::min(forestdist[i + 1][prev_j + 1] + _config->get_insert_cost(),
+                        forestdist[prev_i + 1][prev_j + 1] + 
+                        (_post_order_src[i] == _post_order_dst[j] ? 0 : _config->get_relabel_cost())));
+                treedist[i][j] = forestdist[i + 1][j + 1];
+                computed[i][j] = true;
```

Comment:
This is some cool work.
While I understand why this ```computed``` is here, I am also thinking wouldn't it be better to have either an optional<unsigned int> as treedist[i][j] ? 
Or even a -1 special value to mark them as not yet computed ? From my understanding of that paper, the distance is always bigger than 0 so -1 seems a safe alternative, if not, ```std::optional``` is even better.

Keeping matrices in sync is quire error prone. If it's None, then it needs to be computed, else we are ok to return whatever is there. In this way there's no need to also keep the computed parameter.